[TLI] Add getLibFunc that accepts an Opcode and scalar Type.

paschalis-mpeis opened the pull request:
It sets a LibFunc similarly with the other two getLibFunc methods. Currently, it supports only the FRem Instruction.

Add tests for FRem.

# Motivation:
Almost all libm functions are represented in LLVM as intrinsics with the exception of `fmod` that is transformed into the `FREM` Instruction. This difference means existing code to map scalar calls to vector functions (as used by `LoopVectorize` and `ReplaceWithVecLib`) does not work and so first we want to map the instruction back to it’s representative scalar function. 

Given variants of getLibFunc exist to map an intrinsic to such functions it seems reasonable to have a variant for instructions as well.

paschalis-mpeis committed:
[TLI] Add getLibFunc in API that accepts an Instruction.

It sets a LibFunc similarly with the other two getLibFunc methods.
Currently, it supports only the FRem Instruction.

Add tests for FRem.

nikic commented:
What's the motivation for this?

paulwalker-arm commented on llvm/lib/Analysis/TargetLibraryInfo.cpp:
```diff
@@ -1149,6 +1149,22 @@ bool TargetLibraryInfoImpl::getLibFunc(const Function &FDecl,
+    F = LibFunc_fmod;
+  else if (ScalarTy->isFloatTy())
+    F = LibFunc_fmodf;
+  else if (ScalarTy->isFP128Ty())
+    F = LibFunc_fmodl;
```
This is not strictly speaking true because the choice of what underlying LLVM type is used to represent a long double is target specific.  I think it's safer to drop this code.

paschalis-mpeis committed:
Dropping FP128 as they are target specific.

paschalis-mpeis commented:
> What's the motivation for this?

Most all libm functions are represented in LLVM as intrinsics with the exception of fmod that is transformed into the FREM instruction. This difference means existing code to map scalar calls to vector functions (as used by LoopVectorize and ReplaceWithVecLib) does not work and so first we want to map the instruction back to it’s representative scalar function. Given variants of getLibFunc exist to map an intrinsic to such functions it seems reasonable to have a variant for instructions as well.

labrinea commented on llvm/lib/Analysis/TargetLibraryInfo.cpp:
```diff
@@ -1149,6 +1149,20 @@ bool TargetLibraryInfoImpl::getLibFunc(const Function &FDecl,
+bool TargetLibraryInfoImpl::getLibFunc(const Instruction &I, LibFunc &F) const {
+  if (I.getOpcode() != Instruction::FRem)
+    return false;
+
```
I would write it as
```
Type *ScalarTy = I.getType()->getScalarType();

if (!ScalarTy->isDoubleTy() && !ScalarTy->isFloatTy())
  return false;

F = ScalarTy->isDoubleTy() ? LibFunc_fmod : LibFunc_fmodf;
return true;
```

paulwalker-arm commented on llvm/unittests/Analysis/TargetLibraryInfoTest.cpp:
```diff
@@ -621,3 +622,62 @@ TEST_F(TargetLibraryInfoTest, ValidProto) {
+
+namespace {
+
+// Creates TLI for AArch64 and VecLibrary ARmPL, and uses it to get the TLI
```
Typo "ARmPL"

That said, what's the relevance to a vector library given the test only relates to scalar functions?

paulwalker-arm commented on llvm/unittests/Analysis/TargetLibraryInfoTest.cpp:
```diff
@@ -621,3 +622,62 @@ TEST_F(TargetLibraryInfoTest, ValidProto) {
+  }
+
+  /// Creates an FRem Instruction of Type \p Ty, and uses it to get the TLI
+  /// function name.
+  StringRef getFremScalarName(Type *Ty) {
+    // Use a dummy function and a BB to create an FRem Instruction.
+    FunctionType *FTy = FunctionType::get(Ty, {Ty, Ty}, false);
+    Function *F = Function::Create(FTy, Function::ExternalLinkage, "foo", *M);
+    BasicBlock *BB = BasicBlock::Create(Ctx, "entry", F);
+    IRBuilder<> Builder(BB);
+    Builder.SetInsertPoint(BB);
+    auto *FRem =
+        dyn_cast<Instruction>(Builder.CreateFRem(F->getArg(0), F->getArg(1)));
+
```
Is this necessary.  Does creating the instruction directly (e.g. `BinaryOperator::Create()`) not work?

paschalis-mpeis committed:
Simplified code.

Test class:
- no longer using IRBuilder to create FRem instructions
- no need to use ArmPL vector library

paulwalker-arm commented on llvm/lib/Analysis/TargetLibraryInfo.cpp:
```diff
@@ -1149,6 +1149,18 @@ bool TargetLibraryInfoImpl::getLibFunc(const Function &FDecl,
+  if (I.getOpcode() != Instruction::FRem)
+    return false;
+
+  Type *ScalarTy = I.getType()->getScalarType();
```
I must have missed this before but this should just be `I.getType();` because otherwise you'll return a value for vector types, which would be wrong.

paulwalker-arm commented on llvm/unittests/Analysis/TargetLibraryInfoTest.cpp:
```diff
@@ -621,3 +621,44 @@ TEST_F(TargetLibraryInfoTest, ValidProto) {
+  TLITestAarch64() : TargetTriple(Triple("aarch64-unknown-linux-gnu")) {
+    TLII = std::make_unique<TargetLibraryInfoImpl>(
+        TargetLibraryInfoImpl(TargetTriple));
+    TLII->addVectorizableFunctionsFromVecLib(TargetLibraryInfoImpl::NoLibrary,
+                                             TargetTriple);
```
Is this necessary? I would have thought TLI can exist without needing to add vector mappings.

paschalis-mpeis committed:
getLibFunc accepts an Opcode and a ScalarFTy.

paschalis-mpeis commented:
Addressed review with 24617b5 and rebased to main.

paulwalker-arm commented on llvm/unittests/Analysis/TargetLibraryInfoTest.cpp:
```diff
@@ -621,3 +621,42 @@ TEST_F(TargetLibraryInfoTest, ValidProto) {
+  /// function name.
+  StringRef getFremScalarName(Type *Ty) {
+    Value *V = Constant::getNullValue(Ty);
+    Instruction *FRem = BinaryOperator::Create(Instruction::FRem, V, V);
```
With the new interface you don't need to create the instruction anymore because you can just pass in `Instruction::FRem` and `Ty`.

paulwalker-arm commented on llvm/include/llvm/Analysis/TargetLibraryInfo.h:
```diff
@@ -156,6 +156,10 @@ class TargetLibraryInfoImpl {
 
+  /// Searches for a function name using an Instruction \p Opcode.
+  /// Currently, only the frem instruction is supported.
+  bool getLibFunc(unsigned int Opcode, Type *ScalarTy, LibFunc &F) const;
```
I'd just use `Ty` instead of `ScalarTy` because the interface doesn't care what the type is.

paulwalker-arm reviewed (approved):
A couple of minor things but otherwise looks good to me.

paschalis-mpeis committed:
Renamed ScalarTy to Ty in getLibFunc.

In test no longer creating an FRem instruction.
